src/alpaka/AlpakaCore/alpakaWorkDivHelper.h

```diff
@@ -320,6 +320,474 @@ namespace cms {
+        const Vec3D gridDimension(alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc));
+
+        thread_ = {blockDimension[0u] * blockIdxInGrid[0u] + threadIdxLocal[0u],
+                   blockDimension[1u] * blockIdxInGrid[1u] + threadIdxLocal[1u],
+                   blockDimension[2u] * blockIdxInGrid[2u] + threadIdxLocal[2u]};
+        stride_ = {blockDimension[0u] * gridDimension[0u], blockDimension[1u] * gridDimension[1u], Idx{1}};
+        extent_ = stride_;
+
+        blockDim = blockDimension;
+      }
+
+      ALPAKA_FN_ACC elements_with_stride_2d(const T_Acc& acc, Vec3D extent, Vec3D elementIdxShift = Vec3D::all(0))
+          : extent_(extent + elementIdxShift) {
+        const Vec3D threadIdxLocal(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc));
+        const Vec3D blockIdxInGrid(alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc));
+
+        const Vec3D blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc));
+        const Vec3D gridDimension(alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc));
+
+        thread_ = {blockDimension[0u] * blockIdxInGrid[0u] + threadIdxLocal[0u],
+                   blockDimension[1u] * blockIdxInGrid[1u] + threadIdxLocal[1u],
+                   blockDimension[2u] * blockIdxInGrid[2u] + threadIdxLocal[2u]};
+        thread_ = thread_ + elementIdxShift;
+        stride_ = {blockDimension[0u] * gridDimension[0u], blockDimension[1u] * gridDimension[1u], Idx{1}};
+
+        blockDim = blockDimension;
+      }
+
+      class iterator {
+        friend class elements_with_stride_2d;
+
+      public:
+        ALPAKA_FN_ACC Vec3D operator*() const { return index_; }
+
+        ALPAKA_FN_ACC constexpr iterator& operator++() {
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+          // increment the first coordinate
+          index_[0u] += stride_[0u];
+          if (index_[0u] < extent_[0u])
+            return *this;
+
+          // if the first coordinate overflows, reset it and increment the second
+          // coordinate
+          index_[0u] = thread_[0u];
+          index_[1u] += stride_[1u];
+          if (index_[1u] < extent_[1u])
+            return *this;
+#else
+          // increment the 3rd index and check its value
+          index_[2u] += 1;
+          if (index_[2u] == old_index_[2u] + blockDim[2u])
+            index_[2u] = old_index_[2u];
+
+          //  if the 3rd index was reset, increment the 2nd index
+          if (index_[2u] == old_index_[2u])
+            index_[1u] += 1;
+          if (index_[1u] == old_index_[1u] + blockDim[1u] || index_[1u] == extent_[1u])
+            index_[1u] = old_index_[1u];
+
+          // if the 3rd and 2nd indices were set, increment the first coordinate
+          if (index_[1u] == old_index_[1u] && index_[2u] == old_index_[2u])
+            index_[0u] += 1;
+
+          if (index_[0u] < old_index_[0u] + blockDim[0u] && index_[0u] < extent_[0u] && index_[1u] < extent_[1u]) {
+            return *this;
+          }
```

Comment:
This part seems inconsistent with the `ALPAKA_ACC_GPU_CUDA_ENABLED` case above: there the iteration is only over the 0th and 1st indices, here is over all three indices.